libraries/standard/mqtt/src/mqtt_state.c

```diff
@@ -0,0 +1,611 @@
+                                 packetId,
+                                 &tempQoS,
+                                 &tempState );
+
+    if( qos == MQTTQoS0 )
+    {
+        status = MQTTSuccess;
+    }
+    else if( packetId == MQTT_PACKET_ID_INVALID )
+    {
+        status = MQTTBadParameter;
+    }
+    /* Make sure there's no collision. */
+    else if( recordIndex >= MQTT_STATE_ARRAY_MAX_COUNT )
+    {
+        status = _addRecord( pMqttContext->outgoingPublishRecords,
+                             MQTT_STATE_ARRAY_MAX_COUNT,
+                             packetId,
+                             qos,
+                             MQTTPublishSend );
+    }
+    else
+    {
+        /* If the ID was found there must be a collision. */
+        status = MQTTStateCollision;
+    }
+    
+    return status;
+}
+
+MQTTPublishState_t MQTT_CalculateStatePublish( MQTTStateOperation_t opType,
+                                               MQTTQoS_t qos )
+{
+    MQTTPublishState_t calculatedState = MQTTStateNull;
+    switch( qos )
+    {
+        case MQTTQoS0:
+            calculatedState = MQTTPublishDone;
+            break;
+        case MQTTQoS1:
+            calculatedState = ( opType == MQTT_SEND )? MQTTPubAckPending : MQTTPubAckSend;
+            break;
+        case MQTTQoS2:
+            calculatedState = ( opType == MQTT_SEND )? MQTTPubRecPending : MQTTPubRecSend;
+            break;
+        default:
+            /* No other QoS values. */
+            break;
+    }
+    return calculatedState;
+}
+
+MQTTPublishState_t MQTT_UpdateStatePublish( MQTTContext_t * pMqttContext,
+                                            uint16_t packetId,
+                                            MQTTStateOperation_t opType,
+                                            MQTTQoS_t qos )
+{
+    MQTTPublishState_t newState = MQTTStateNull;
+    MQTTPublishState_t currentState = MQTTStateNull;
+    MQTTStatus_t mqttStatus = MQTTSuccess;
+    size_t recordIndex = MQTT_STATE_ARRAY_MAX_COUNT;
+    MQTTQoS_t foundQoS = MQTTQoS0;
+    MQTTPubAckInfo_t * records = NULL;
+    bool isTransitionValid = false;
+
+    /* Set this pointer so some common code can be used later. */
+    if( opType == MQTT_SEND )
+    {
+        records = pMqttContext->outgoingPublishRecords;
+    }
+    else
+    {
+        records = pMqttContext->incomingPublishRecords;
+    }
+
+    recordIndex = _findInRecord( records,
+                                 MQTT_STATE_ARRAY_MAX_COUNT,
+                                 packetId,
+                                 &foundQoS,
+                                 &currentState );
+    
+    if( qos == MQTTQoS0 )
+    {
+        /* QoS 0 publish. Do nothing */
+        newState = MQTTPublishDone;
+    }
+    else if( packetId == MQTT_PACKET_ID_INVALID )
+    {
+        /* Publishes > QoS 0 need a valid packet ID. */
+        mqttStatus = MQTTBadParameter;
+    }
+    else if( foundQoS == MQTTQoS0 )
```

Comment:
Is it possible ? Should this be an assert()

Reply:
Yes, it means the ID was not found in the record, so `_findInRecord` was not able to update `foundQoS`.

Reply:
Earlier on line 341, it is checked by condition: if( recordIndex >= MQTT_STATE_ARRAY_MAX_COUNT ) ,  if( foundQoS == MQTTQoS0 ) looks bit confusing.

Reply:
They're equivalent for this purpose, since both indicate that the record was not found. I chose to use `foundQoS` since that's the variable that gets updated in the block. Do you think it should be changed?

Reply:
Refactored